Add vga driver and temporal GUI

jerexio opened the pull request:
Resolved [zrafa](https://github.com/zrafa)/[xinu-x86-gui](https://github.com/zrafa/xinu-x86-gui) #1 

Two commits are attached:

1.  Add VGA driver, modify init, seek, and write. Also, modify the Config file to enable VGA open.
2. Delete the vga.c and text_on_vga.c files from ./lib. Add a temporary GUI to ./system, which unifies deleted files.

I created the driver trying to follow the idea I read here: [Vga Tasks](https://se.fi.uncoma.edu.ar/xinu-x86-gui/doc/vga_task.txt)

I extracted the functions from the vgac and text_on_vga.c files. This was done to unify the GUI into a single file.
Now, regarding the question "Why is it in the system folder?" This is because it was the only directory where it made sense. It seemed strange to me that a file from ./lib would use device functions (such as open, write, etc.).

zrafa commented:
Hi Jere,
I cc to xinu-ui-project mailing list.

About your patches: Great work!!! That is a very good work!.
I have a couple of quick comments before merge:

1- About temporary gui code under system/
I understand your comments. However, system/ is the folder for the
core kernel components,
and those functions (set_pixel(), text_on_vga, etc) are definitely
library functions.
In order to reply your comments, you can check some functions under lib/, and
you will see  that some library functions call to syscalls or kernel components
(for example fput() underl lib/ calls to syscall putc()). That is okey
and the proper way.
If you want to understand further, try to think how microui will use
the vga driver:
open(VGA);
gui_put_surface(x, y, surface);  /* that function does not exist, it is an
                                                      * example.
surface is a memory buffer with
                                                      * some image,
which user wants to put on screen on (x,y).
                                                      * Then
gui_put_surface will call to low level syscalls
                                                      * (probably
seek(VGA, y*width*(bpp/4); write(VGA, ...))
                                                      * several times,
one seek/write per line of surface
                                                      */

2- about some functions in temporary gui: most of them are probably
old drafts. You can remove all the functions which are never called.

3- setPixel(): I would change the name to gui_set_pixel()
(also, the same for the others: gui_something). All those names are
ancient drafts as well. Maybe that utils vga lib could be named simply lib/gui.c

4- in vgawrite.c:
The code is okey I think, but maybe you can use simply memcpy()?
(not sure)

5- for(uint32 i = 0; i < count; i += sizeof(uint32)){
uint32 *pixel = fb;

Better coding style is to declare local vars at the beggining of the
funcion, before any c sentence.
For example:
devcall vgawrite (
        struct dentry   *devptr,        /* Entry in device switch table */
        char *buffer,        /* Address of buffer            */
        uint32        count  /* Length of buffer             */
        )
{
    void *fb;
    uint32 i;
    uint32 *pixel;
    /* one blank line space */
    /* first c sentence */

Re-work whenever you want a bit the code. Ask any questions you have, and
send your new version of patches.
Again, cool work (it is cool because we need a proper vga driver before to
port microui).

Rafa



On Sat, Jun 17, 2023 at 8:21 PM Jeremias Rodriguez
***@***.***> wrote:
>
> Two commits are attached:
>
> Add VGA driver, modify init, seek, and write. Also, modify the Config file to enable VGA open.
> Delete the vga.c and text_on_vga.c files from ./lib. Add a temporary GUI to ./system, which unifies deleted files.
>
> I created the driver trying to follow the idea I read here: Vga Tasks
>
> I extracted the functions from the vgac and text_on_vga.c files. This was done to unify the GUI into a single file.
> Now, regarding the question "Why is it in the system folder?" This is because it was the only directory where it made sense. It seemed strange to me that a file from ./lib would use device functions (such as open, write, etc.).
>
> ________________________________
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/zrafa/xinu-x86-gui/pull/2
>
> Commit Summary
>
> c3330de Add vga driver, modify init, seek and write. Also, modify the Config file to enable VGA open.
> 8e3d77a Delete the vga.c and text_on_vga.c files from ./lib . Add a temporary GUI to ./system, which unifies deleted files.
>
> File Changes (14 files)
>
> M config/Configuration (2)
> M device/vga/vgainit.c (14)
> M device/vga/vgaseek.c (2)
> M device/vga/vgawrite.c (22)
> M include/prototypes.h (7)
> M include/vga.h (2)
> D lib/text_on_vga.c (83)
> D lib/vga.c (108)
> M shell/cmdtab.c (2)
> M shell/galaga/galaga.c (1)
> M shell/keyboard.c (3)
> M shell/mouse.c (13)
> M system/main.c (8)
> A system/temporary_gui.c (161)
>
> Patch Links:
>
> https://github.com/zrafa/xinu-x86-gui/pull/2.patch
> https://github.com/zrafa/xinu-x86-gui/pull/2.diff
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

jerexio commented:
Hi Rafa, I updated the write driver using memcpy, and it really became simpler.
Also, I changed the GUI and moved it back to lib/. Regarding what you mentioned about some functions not being called, I checked it, and all of them are used.
In the particular case of gui_set_pixel and gui_pixel, they are identical except that one takes a 16-bit color and the other takes a 32-bit color. I decided to keep both of them (although they could be unified) since it's a temporary GUI.

There's a commit titled "Solve problems to compile" that I uploaded because there were issues with compilation. Maybe we need to remove it.


> Hi Jere, I cc to xinu-ui-project mailing list. About your patches: Great work!!! That is a very good work!. I have a couple of quick comments before merge: 1- About temporary gui code under system/ I understand your comments. However, system/ is the folder for the core kernel components, and those functions (set_pixel(), text_on_vga, etc) are definitely library functions. In order to reply your comments, you can check some functions under lib/, and you will see that some library functions call to syscalls or kernel components (for example fput() underl lib/ calls to syscall putc()). That is okey and the proper way. If you want to understand further, try to think how microui will use the vga driver: open(VGA); gui_put_surface(x, y, surface); /* that function does not exist, it is an * example. surface is a memory buffer with * some image, which user wants to put on screen on (x,y). * Then gui_put_surface will call to low level syscalls * (probably seek(VGA, y*width*(bpp/4); write(VGA, ...)) * several times, one seek/write per line of surface */ 2- about some functions in temporary gui: most of them are probably old drafts. You can remove all the functions which are never called. 3- setPixel(): I would change the name to gui_set_pixel() (also, the same for the others: gui_something). All those names are ancient drafts as well. Maybe that utils vga lib could be named simply lib/gui.c 4- in vgawrite.c: The code is okey I think, but maybe you can use simply memcpy()? (not sure) 5- for(uint32 i = 0; i < count; i += sizeof(uint32)){ uint32 *pixel = fb; Better coding style is to declare local vars at the beggining of the funcion, before any c sentence. For example: devcall vgawrite ( struct dentry *devptr, /* Entry in device switch table */ char *buffer, /* Address of buffer */ uint32 count /* Length of buffer */ ) { void *fb; uint32 i; uint32 *pixel; /* one blank line space */ /* first c sentence */ Re-work whenever you want a bit the code. Ask any questions you have, and send your new version of patches. Again, cool work (it is cool because we need a proper vga driver before to port microui). Rafa
> […](#)
> On Sat, Jun 17, 2023 at 8:21 PM Jeremias Rodriguez ***@***.***> wrote: Two commits are attached: Add VGA driver, modify init, seek, and write. Also, modify the Config file to enable VGA open. Delete the vga.c and text_on_vga.c files from ./lib. Add a temporary GUI to ./system, which unifies deleted files. I created the driver trying to follow the idea I read here: Vga Tasks I extracted the functions from the vgac and text_on_vga.c files. This was done to unify the GUI into a single file. Now, regarding the question "Why is it in the system folder?" This is because it was the only directory where it made sense. It seemed strange to me that a file from ./lib would use device functions (such as open, write, etc.). ________________________________ You can view, comment on, or merge this pull request online at: #2 Commit Summary [c3330de](https://github.com/zrafa/xinu-x86-gui/commit/c3330de8081fb28106b949bff7457c419892c8a5) Add vga driver, modify init, seek and write. Also, modify the Config file to enable VGA open. [8e3d77a](https://github.com/zrafa/xinu-x86-gui/commit/8e3d77ada7e334062027ffe4763a6f47a184dea8) Delete the vga.c and text_on_vga.c files from ./lib . Add a temporary GUI to ./system, which unifies deleted files. File Changes (14 files) M config/Configuration (2) M device/vga/vgainit.c (14) M device/vga/vgaseek.c (2) M device/vga/vgawrite.c (22) M include/prototypes.h (7) M include/vga.h (2) D lib/text_on_vga.c (83) D lib/vga.c (108) M shell/cmdtab.c (2) M shell/galaga/galaga.c (1) M shell/keyboard.c (3) M shell/mouse.c (13) M system/main.c (8) A system/temporary_gui.c (161) Patch Links: https://github.com/zrafa/xinu-x86-gui/pull/2.patch https://github.com/zrafa/xinu-x86-gui/pull/2.diff — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

zrafa commented:
Hi Jere,

On Mon, Jun 19, 2023 at 1:34 PM Jeremias Rodriguez
***@***.***> wrote:
>
> Hi Rafa, I updated the write driver using memcpy, and it really became simpler.

Looks OK!!. I will merge the pull request.
I have several code clean ups and a few adds to the lib/gui.c mainly
in mind, but we can do it on further patches.

> Also, I changed the GUI and moved it back to lib/. Regarding what you mentioned about some functions not being called, I checked it, and all of them are used.

Perfect.

> In the particular case of gui_set_pixel and gui_pixel, they are identical except that one takes a 16-bit color and the other takes a 32-bit color. I decided to keep both of them (although they could be unified) since it's a temporary GUI.

Yes, we will merge them later.

> There's a commit titled "Solve problems to compile" that I uploaded because there were issues with compilation. Maybe we need to remove it.

Yes, that is weird. Like the config/conf.h file says:

/* conf.c (GENERATED FILE; DO NOT EDIT) */

we should not edit it.
I will merge anyway, because it will not cause any trouble. But, we
can check the compilation errors you got/get so we can fix them
properly later.

Rafa

>
> Hi Jere, I cc to xinu-ui-project mailing list. About your patches: Great work!!! That is a very good work!. I have a couple of quick comments before merge: 1- About temporary gui code under system/ I understand your comments. However, system/ is the folder for the core kernel components, and those functions (set_pixel(), text_on_vga, etc) are definitely library functions. In order to reply your comments, you can check some functions under lib/, and you will see that some library functions call to syscalls or kernel components (for example fput() underl lib/ calls to syscall putc()). That is okey and the proper way. If you want to understand further, try to think how microui will use the vga driver: open(VGA); gui_put_surface(x, y, surface); /* that function does not exist, it is an * example. surface is a memory buffer with * some image, which user wants to put on screen on (x,y). * Then gui_put_surface will call to low level syscalls * (probably seek(VGA, ywidth(bpp/4); write(VGA, ...)) * several times, one seek/write per line of surface / 2- about some functions in temporary gui: most of them are probably old drafts. You can remove all the functions which are never called. 3- setPixel(): I would change the name to gui_set_pixel() (also, the same for the others: gui_something). All those names are ancient drafts as well. Maybe that utils vga lib could be named simply lib/gui.c 4- in vgawrite.c: The code is okey I think, but maybe you can use simply memcpy()? (not sure) 5- for(uint32 i = 0; i < count; i += sizeof(uint32)){ uint32 pixel = fb; Better coding style is to declare local vars at the beggining of the funcion, before any c sentence. For example: devcall vgawrite ( struct dentry devptr, / Entry in device switch table / char buffer, / Address of buffer / uint32 count / Length of buffer / ) { void fb; uint32 i; uint32 pixel; / one blank line space / / first c sentence / Re-work whenever you want a bit the code. Ask any questions you have, and send your new version of patches. Again, cool work (it is cool because we need a proper vga driver before to port microui). Rafa
> …
> On Sat, Jun 17, 2023 at 8:21 PM Jeremias Rodriguez @.> wrote: Two commits are attached: Add VGA driver, modify init, seek, and write. Also, modify the Config file to enable VGA open. Delete the vga.c and text_on_vga.c files from ./lib. Add a temporary GUI to ./system, which unifies deleted files. I created the driver trying to follow the idea I read here: Vga Tasks I extracted the functions from the vgac and text_on_vga.c files. This was done to unify the GUI into a single file. Now, regarding the question "Why is it in the system folder?" This is because it was the only directory where it made sense. It seemed strange to me that a file from ./lib would use device functions (such as open, write, etc.). ________________________________ You can view, comment on, or merge this pull request online at: #2 Commit Summary c3330de Add vga driver, modify init, seek and write. Also, modify the Config file to enable VGA open. 8e3d77a Delete the vga.c and text_on_vga.c files from ./lib . Add a temporary GUI to ./system, which unifies deleted files. File Changes (14 files) M config/Configuration (2) M device/vga/vgainit.c (14) M device/vga/vgaseek.c (2) M device/vga/vgawrite.c (22) M include/prototypes.h (7) M include/vga.h (2) D lib/text_on_vga.c (83) D lib/vga.c (108) M shell/cmdtab.c (2) M shell/galaga/galaga.c (1) M shell/keyboard.c (3) M shell/mouse.c (13) M system/main.c (8) A system/temporary_gui.c (161) Patch Links: https://github.com/zrafa/xinu-x86-gui/pull/2.patch https://github.com/zrafa/xinu-x86-gui/pull/2.diff — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you are subscribed to this thread.Message ID: @.>
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>
